components/openvsx-proxy/pkg/run.go

```diff
@@ -112,3 +113,69 @@ func (t *DurationTrackingTransport) RoundTrip(r *http.Request) (*http.Response,
 	}(start)
 	return http.DefaultTransport.RoundTrip(r)
 }
+
+// From go/src/net/http/httputil/reverseproxy.go
+
+func singleJoiningSlash(a, b string) string {
+	aslash := strings.HasSuffix(a, "/")
+	bslash := strings.HasPrefix(b, "/")
+	switch {
+	case aslash && bslash:
+		return a + b[1:]
+	case !aslash && !bslash:
+		return a + "/" + b
+	}
+	return a + b
+}
+
+func joinURLPath(a, b *url.URL) (path, rawpath string) {
+	if a.RawPath == "" && b.RawPath == "" {
+		return singleJoiningSlash(a.Path, b.Path), ""
+	}
+	// Same as singleJoiningSlash, but uses EscapedPath to determine
+	// whether a slash should be added
+	apath := a.EscapedPath()
+	bpath := b.EscapedPath()
+
+	aslash := strings.HasSuffix(apath, "/")
+	bslash := strings.HasPrefix(bpath, "/")
+
+	switch {
+	case aslash && bslash:
+		return a.Path + b.Path[1:], apath + bpath[1:]
+	case !aslash && !bslash:
+		return a.Path + "/" + b.Path, apath + "/" + bpath
+	}
+	return a.Path + b.Path, apath + bpath
+}
+
+func newSingleHostReverseProxy(target *url.URL) *httputil.ReverseProxy {
+	targetQuery := target.RawQuery
+	director := func(req *http.Request) {
+		originalHost := req.Host
+
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+		req.URL.Path, req.URL.RawPath = joinURLPath(target, req.URL)
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+		req.Host = target.Host
+
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+
+		req.Header.Set("X-Forwarded-Host", originalHost)
```

Comment:
What if these headers already present ie someone using Gitpod behind another proxy. Should not we preserve them then?

Reply:
Added logic to preserve headers if they are present but from logs the headers are already present when the request reaches openvsx-proxy in the preview environment so there's already a proxy in front :smile: 
```
{"func":"request_handler","level":"info","message":"handling request","request":"\"POST /vscode/gallery/extensionquery HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: application/json;api-version=3.0-preview.1\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nCache-Control: no-cache\\r\\nContent-Length: 236\\r\\nContent-Type: application/json\\r\\nOrigin: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nPragma: no-cache\\r\\nReferer: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com/\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: empty\\r\\nSec-Fetch-Mode: cors\\r\\nSec-Fetch-Site: same-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nVscode-Sessionid: 37f5a423-2e86-4705-af53-fa9b9c3c461e1660244402281\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\nX-Market-Client-Id: VSCode 1.70.0\\r\\nX-Market-User-Id: 70064d9c-b88e-4cdc-bb46-a2d0e7f5d4cb\\r\\n\\r\\n\"","request_content_length":"236","request_id":"060cc98b-7dfb-4e55-a3db-d1e699ec36a5","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":947558,"duration_human":"947.558µs","func":"request_handler","level":"info","message":"processing request finished","request":"POST /vscode/gallery/extensionquery 236 653ddf16318e0343fd552c6dccda25c78e550d3d","request_content_length":"236","request_id":"060cc98b-7dfb-4e55-a3db-d1e699ec36a5","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"request_handler","level":"info","message":"handling request","request":"\"POST /vscode/gallery/extensionquery HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: application/json;api-version=3.0-preview.1\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nCache-Control: no-cache\\r\\nContent-Length: 236\\r\\nContent-Type: application/json\\r\\nOrigin: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nPragma: no-cache\\r\\nReferer: https://gitpodio-gitpodvscodede-vs7zasevumf.ws.jp-openvsx123d49373e.preview.gitpod-dev.com/\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: empty\\r\\nSec-Fetch-Mode: cors\\r\\nSec-Fetch-Site: same-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nVscode-Sessionid: 37f5a423-2e86-4705-af53-fa9b9c3c461e1660244402281\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\nX-Market-Client-Id: VSCode 1.70.0\\r\\nX-Market-User-Id: 70064d9c-b88e-4cdc-bb46-a2d0e7f5d4cb\\r\\n\\r\\n\"","request_content_length":"236","request_id":"f95d2aca-12d7-40fa-95e1-84e60d220a7e","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":583004,"duration_human":"583.004µs","func":"request_handler","level":"info","message":"processing request finished","request":"POST /vscode/gallery/extensionquery 236 653ddf16318e0343fd552c6dccda25c78e550d3d","request_content_length":"236","request_id":"f95d2aca-12d7-40fa-95e1-84e60d220a7e","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"duration":204211504,"duration_human":"204.211504ms","func":"transport_roundtrip","level":"info","message":"upstream call finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"response_handler","level":"info","message":"handling response","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"func":"response_handler","level":"info","message":"successfully stored response to cache","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"duration":1184627,"duration_human":"1.184627ms","func":"response_handler","level":"info","message":"processing response finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","response_content_length":"0","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","status":"302","time":"2022-08-11T19:00:47Z"}
{"duration":209098713,"duration_human":"209.098713ms","func":"request_handler","hit_cache_backup":true,"hit_cache_regular":false,"level":"info","message":"request finished","request":"GET /vscode/asset/redhat/java/1.9.0/Microsoft.VisualStudio.Services.Icons.Default 0 da39a3ee5e6b4b0d3255bfef95601890afd80709","request_content_length":"0","request_id":"d56412da-eda8-42d2-8048-6ac546704eb7","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
{"func":"request_handler","level":"info","message":"handling request","request":"\"GET /vscode/asset/wesleyegberto/vscode-java-tests/1.5.0/Microsoft.VisualStudio.Services.Icons.Default HTTP/1.1\\r\\nHost: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nAccept: image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8\\r\\nAccept-Encoding: gzip, deflate, br\\r\\nAccept-Language: en-US,en;q=0.9\\r\\nSec-Ch-Ua: \\\"Chromium\\\";v=\\\"104\\\", \\\" Not A;Brand\\\";v=\\\"99\\\", \\\"Google Chrome\\\";v=\\\"104\\\"\\r\\nSec-Ch-Ua-Mobile: ?0\\r\\nSec-Ch-Ua-Platform: \\\"Linux\\\"\\r\\nSec-Fetch-Dest: image\\r\\nSec-Fetch-Mode: no-cors\\r\\nSec-Fetch-Site: cross-site\\r\\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36\\r\\nX-Forwarded-For: 10.0.2.2\\r\\nX-Forwarded-Host: open-vsx.jp-openvsx123d49373e.preview.gitpod-dev.com\\r\\nX-Forwarded-Proto: https\\r\\n\\r\\n\"","request_content_length":"0","request_id":"30ecf423-2fd5-45bb-884a-91cd107bab87","serviceContext":{"service":"openvsx-proxy","version":""},"severity":"INFO","time":"2022-08-11T19:00:47Z"}
```